### PR TITLE
fix: more flexible polymorphic types lookup

### PR DESCRIPTION
**Pull request by bf4**
Includes:
- fix: more flexible polymorphic types lookup
- refactor: lookup polymorphic types only once

### All Submissions:

- [ ] I've checked to ensure there aren't other open [Pull Requests](https://github.com/cerebris/jsonapi-resources/pulls) for the same update/change.
- [ ] I've submitted a [ticket](https://github.com/cerebris/jsonapi-resources/issues) for my issue if one did not already exist.
- [ ] My submission passes all tests. (Please run the full test suite locally to cut down on noise from travis failures.)
- [ ] I've used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message or the description.
- [ ] I've added/updated tests for this change.

### New Feature Submissions:

- [ ] I've submitted an issue that describes this feature, and received the go ahead from the maintainers.
- [ ] My submission includes new tests.
- [ ] My submission maintains compliance with [JSON:API](http://jsonapi.org/).

### Bug fixes and Changes to Core Features:

- [ ] I've included an explanation of what the changes do and why I'd like you to include them.
- [ ] I've provided test(s) that fails without the change.

### Reviewer Checklist:
- [ ] Maintains compliance with JSON:API
- [ ] Adequate test coverage exists to prevent regressions

### CODE REVIEW

File: lib/jsonapi/relationship.rb [hunk 1]
```diff
@@ -90,14 +90,37 @@ def self.polymorphic_types(name)
       @poly_hash ||= {}.tap do |hash|
         ObjectSpace.each_object do |klass|
           next unless Module === klass
-          if ActiveRecord::Base > klass
+          is_active_record_inspectable = ActiveRecord::Base > klass
+          is_active_record_inspectable &&= klass.respond_to?(:reflect_on_all_associations, true)
```
Comment (bf4) [inline lib/jsonapi/relationship.rb:94]: @lgebhardt I'll add a test for this. I'm testing against our app on this branch right now and looks good

File: lib/jsonapi/resource_common.rb [hunk 2]
```diff
@@ -992,17 +992,7 @@ def polymorphic(polymorphic = true)
...
-            next unless Module === klass
-            if klass < ActiveRecord::Base
-              klass.reflect_on_all_associations(:has_many).select{|r| r.options[:as] }.each do |reflection|
-                (hash[reflection.options[:as]] ||= []) << klass.name.underscore
-              end
-            end
-          end
-        end
-        @poly_hash[_polymorphic_name.to_sym]
+        JSONAPI::Relationship.polymorphic_types(_polymorphic_name.to_sym)
```
Comment (bf4) [inline lib/jsonapi/resource_common.rb:1023]: @lgebhardt is a refactor included in the PR as a separate commit

File: lib/jsonapi/relationship.rb [hunk 3]
```diff
@@ -90,14 +90,37 @@ def self.polymorphic_types(name)
...
             klass.reflect_on_all_associations(:has_many).select { |r| r.options[:as] }.each do |reflection|
-              (hash[reflection.options[:as]] ||= []) << klass.name.underscore
+              (hash[reflection.options[:as]] ||= []) << format_polymorphic_klass_type(klass).underscore
             end
           end
         end
       end
-      @poly_hash[name.to_sym]
+      @poly_hash.fetch(name.to_sym) do
+        klass = name.classify.safe_constantize
```
Comment (bf4) [inline lib/jsonapi/relationship.rb:104]: fallback strategy I'm trying

**Comment by lgebhardt** [comment]
Supplanted by #1433

**Pull request closed without merging**